### PR TITLE
JDK-8277520: Implement JDK-8 default methods for IdentityHashMap

### PR DESCRIPTION
**Pull request by liach**
Might need a CSR as now `computeIfAbsent` `computeIfPresent` `compute` `merge` would throw CME if the functions modified the map itself, and there are corresponding specification changes.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request to be approved

### Issues
 * [JDK-8277520](https://bugs.openjdk.org/browse/JDK-8277520): Implement JDK-8 default methods for IdentityHashMap
 * [JDK-8278313](https://bugs.openjdk.org/browse/JDK-8278313): Implement JDK-8 default methods for IdentityHashMap (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6532/head:pull/6532` \
`$ git checkout pull/6532`

Update a local copy of the PR: \
`$ git checkout pull/6532` \
`$ git pull https://git.openjdk.java.net/jdk pull/6532/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6532`

View PR using the GUI difftool: \
`$ git pr show -t 6532`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6532.diff">https://git.openjdk.java.net/jdk/pull/6532.diff</a>

</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
Yes, behavioral changes in a widely-used class would merit CSR review.

**Comment by liach** [comment]
Since I don't have an account on the JBS, I will post the CSR markdown draft here and hope someone can create one for me.
```md
CSR of: JDK-8277520

Component: core-libs
Subcomponent: java.util:collections
Compatibility Kind: behavioral
Compatibility risk: low
Compatibility Risk Description: Previously successful calls to `IdentityHashMap::computeIfAbsent` etc. may now fail.
Interface Kind: Java API
Scope: SE
Description:

## Summary
Explicitly state that the `computeIfAbsent`, `computeIfPresent`, `compute`, and `merge` methods in `IdentityHashMap` will throw `ConcurrentModificationException` when the (re)mapping function modifies the map.

## Problem
Previously, with the default implementation provided by the `Map` interface, the 4 methods in `IdentityHashMap` will never throw `ConcurrentModificationException`, but the efficient override implementations requires the (re)mapping function to not modify this map when (re)mapping function's modification to the map is detected.

## Solution
Adds documentation to specify the new behavior of the 4 methods in `IdentityHashMap`. Such behavior is compatible with that specified in the `Map` superinterface.

## Specification
We propose to add these documentation for the 4 concurrency-hostile methods as follows: (implementations ignored)
```java
    /**
     * {@inheritDoc}
     *
     * <p>This method will, on a best-effort basis, throw a
     * {@link ConcurrentModificationException} if it is detected that the
     * mapping function modifies this map during computation.
     *
     * @throws ConcurrentModificationException if it is detected that the
     * mapping function modified this map
     */
    @Override
    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
    }

    /**
     * {@inheritDoc}
     *
     * <p>This method will, on a best-effort basis, throw a
     * {@link ConcurrentModificationException} if it is detected that the
     * remapping function modifies this map during computation.
     *
     * @throws ConcurrentModificationException if it is detected that the
     * remapping function modified this map
     */
    @Override
    public V computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
    }

    /**
     * {@inheritDoc}
     *
     * <p>This method will, on a best-effort basis, throw a
     * {@link ConcurrentModificationException} if it is detected that the
     * remapping function modifies this map during computation.
     *
     * @throws ConcurrentModificationException if it is detected that the
     * remapping function modified this map
     */
    @Override
    public V compute(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
    }

    /**
     * {@inheritDoc}
     *
     * <p>This method will, on a best-effort basis, throw a
     * {@link ConcurrentModificationException} if it is detected that the
     * remapping function modifies this map during computation.
     *
     * @throws ConcurrentModificationException if it is detected that the
     * remapping function modified this map
     */
    @Override
    public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
    }
```
```

**Comment by liach** [comment]
Just curious, what else is recommended for such additions? Is there any recommended tests (though previous additions of `forEach` and `replaceAll` did not add any)

**Comment by liach** [comment]
Patch still active.

**Comment by RogerRiggs** [comment]
The PR description doesn't do the change request justice. At least the first paragraph from the bug report would be useful.

The rationale seems ok but not compelling: 
"may save a few hash table lookups and moderately boost the performance of IdentityHashMap".

Please quantify the improvement with JMH test results. 
There may be an existing test in `test/micro/org/openjdk/bench/java/util` to build on.

Thanks

**Comment by liach** [comment]
Slightly busy now, will write benchmarks later and publish results for before and after.

**Comment by liach** [comment]
[Baseline results](https://github.com/openjdk/jdk/files/8112532/micro.txt) (See https://github.com/liachmodded/jdk/tree/identityhashmap-bench)

[Result with this patch](https://github.com/openjdk/jdk/files/8112533/micro.txt)

Seems the benchmark does have some problems within itself. maybe need more consistent data sets for more consistent results (test both put and putIfAbsent against the same data set)

**Comment by stuart-marks** [comment]
@liach I don't have much time to spend on this. Several comments.

JMH benchmarking is more than a bit of an art. There's a lot of information in the JMH samples, so I'd recommend going through them in detail if you haven't already. There are a couple of obvious things to look at, such as to make sure to return values produced by the computation (or use black holes); to fork multiple JVMs during the benchmark run; to reduce or eliminate garbage generation during the test, or account for it if it can't be eliminated; and so forth.

In this particular case it's not clear to me how much performance there is to be gained from overriding the default methods. Suppose an entry exists and `compute(k, bifunc)` is called. The default method calls `get` to obtain the value, calls the bifunction, then calls `put(k, newVal)`. An optimized implementation would remember the location of the mapping so that the new value could be stored without probing again. But probing in an IdentityHashMap is likely pretty inexpensive: the identity hashcode is cached, finding the table index is integer arithmetic, and searching for the right mapping is reference comparisons on table entries that are probably already cached. It doesn't seem likely to me that there is much to be gained here in the first place.

Then again, one's intuition about performance, including mine, is easily wrong! But: if you're having trouble writing a benchmark that demonstrates a performance improvement, maybe that means there isn't much performance to be gained.

As a general comment I'd suggest pursuing performance improvements only when there's a demonstrated performance issue. This kind of looks to me like a case of starting with "I bet I can speed this up by changing this code" and then trying to justify that with benchmarks. If so, that's kind of backwards, and it can easily lead to a lot of time being wasted.

**Comment by ExE-Boss** [comment]
@stuart-marks
Note that the default implementation of those methods assume that the map uses `Object.equals(…)` and `Object.hashCode()`, which doesn’t apply to `IdentityHashMap`.

This means that the performance argument is secondary to implementation correctness.

File: src/java.base/share/classes/java/util/IdentityHashMap.java [hunk 1]
```diff
@@ -555,25 +523,14 @@ public V remove(Object key) {
     private boolean removeMapping(Object key, Object value) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/IdentityHashMap.java:523]: Note that `IdentityHashMap.remove(Object, Object)` needs to call this method, as the default implementation calls `Objects.equals(…)`: <https://github.com/openjdk/jdk/blob/3f923b82c31325504430b50dee262fd460004e7b/src/java.base/share/classes/java/util/Map.java#L866-L874>

**Comment by liach** [comment]
> @stuart-marks Note that the default implementation of those methods assume that the map uses `Object.equals(…)` and `Object.hashCode()`, which doesn’t apply to `IdentityHashMap`.
> 
> This means that the performance argument is secondary to implementation correctness.

That was tracked in a separate ticket JDK-8178355. My current implementations utilize the newly added `findInterestingIndex`, which I plan to use in a fix for that bug in addition, as opposed to writing a while loop in virtually every method in the class.

**Comment by liach** [comment]
Also re stuart: The relative performance of `putIfAbsent` versus `put` has always improved with this patch, now being 1.01 times of `put` call time while previously it was 1.2 times of it, regardless of jvm instance noise.

My main concern is the performance hit brought by my move of the while loops in every method into a consolidated `findInterestingIndex`. This part is somehow harder to get consistent results, or should I make a clean copy of `IdentityHashMap` on new implementation and test (works only if there is no existing hotspot optimizations on `IdentityHashMap`)

File: src/java.base/share/classes/java/util/IdentityHashMap.java (hunk 1 above)
Comment (ExE-Boss) [reply src/java.base/share/classes/java/util/IdentityHashMap.java:523]: This is being addressed in <https://github.com/openjdk/jdk/pull/8259>.

**Comment by ExE-Boss** [comment]
There should probably be something like [test/jdk/java/util/Collections/Wrappers.java](https://github.com/openjdk/jdk/blob/master/test/jdk/java/util/Collections/Wrappers.java) to check that `IdentityHashMap` overrides all `default` methods from `java.util.Map` (with `remove(K, V)` and `replace(K, V, V)` depending on <https://github.com/openjdk/jdk/pull/8259>).

**Pull request closed without merging**